---
title: 'SLE712 AT3: BIoinformatics Assignment'
author: "Jessica Silver"
date: 2022-06-03
output: html_document
theme: paper
  
---

source: https://github.com/jessicasilver/sle712_at3

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### Set Up Instructions

Prior to running all commands, please ensure that `seqinr`, `kableExtra` and `R.utils` packages are installed. 

```{r setupinstructions}
suppressPackageStartupMessages({
  library("seqinr")
library("kableExtra")
library("R.utils")
})

```

***

### Part 1
#### Question 1. Read in the file, making the gene identifiers the row names. Show a table of values for the first six genes. 


``` {r gene.expression.tsv-codechunk1}
gene.expression.URL <- "https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/gene_expression.tsv" #link to file
gene.expression.tsv <- read.table(gene.expression.URL,stringsAsFactors = FALSE, header=TRUE, row.names = 1) #row.names = 1 to keep gene identifiers as row names, not integers

gene.expression.tsv[1:6,] %>% #show first 6 genes/rows #pipe to kableExtra
  kbl(caption = "Table 1. Part 1 Question 1: Gene Expression dataset showing values for the first six genes") %>%
  kable_classic(full_width = F, html_font = "Cambria", position = "center", font_size = 12) 
```

***

#### Question 2. Make a new column which is the mean of the other columns. Show a table of values for the first six genes.


``` {r gene.expression.tsv-codechunk2}
gene.expression.tsv$GE.Means <- rowMeans(gene.expression.tsv) #calculate average gene expression between the three samples of interest for all genes.
gene.expression.tsv <- round(gene.expression.tsv, digits = 0) #round to no decimal places as in original dataset

gene.expression.tsv[1:6,] %>% #show first 6 genes/rows #pipe to kableExtra
  kbl(caption = "Table 2. Part 1 Question 2: Gene Expression dataset showing values for the first six genes in each sample, and the mean expression across all samples.") %>%
  kable_classic(full_width = F, html_font = "Cambria", position = "center", font_size = 12) 
```

***

#### Question 3. List the 10 genes with the highest mean expression

``` {r gene.expression.tsv-codechunk3}
reverse_sortedGEMeans <- gene.expression.tsv[order(-gene.expression.tsv$GE.Means),] #order(-df$colname) to sort data from highest to lowest

reverse_sortedGEMeans[1:10,4,drop=FALSE] %>% #drop=FALSE retains the dimension of the object #pipe to kableExtra
  kbl(caption = "Table 3. Part 1 Question 3: Gene Expression dataset showing the 10 genes with the highest mean expression across the three samples.") %>%
  kable_classic(full_width = F, html_font = "Cambria", position = "center", font_size = 12) 
```

***

#### Question 4. Determine the number of genes with a mean < 10


```{r gene.expression.tsv-codechunk4}
print(nrow(gene.expression.tsv[gene.expression.tsv$GE.Means <10,])) #nrow with condition "number genes with a mean < 10"
```

***

#### Question 5. Make a histogram plot of the mean values and include it into your report.

```{r gene.expression.tsv-codechunk5A}
hist(gene.expression.tsv$GE.Means,main="Part-1 Q-5: Mean Gene Expression Values", xlab="Mean Gene Expression",breaks="scott",xlim=c(0,10000),ylim = c(0,50000))
```

**Figure 1A. ** Part 1 Question 5: Histogram showing the frequency of mean gene expression from the "gene expression" dataset. 

Figure 1A shows over 40,000 genes that have a mean expression less than 100. 
For greater resolution of genes that have a mean expression of over 2000, the x- and y-axis limits were adjusted. 

```{r gene.expression.tsv-codechunk5B}
hist(gene.expression.tsv$GE.Means,main="Part-1 Q-5: Mean Gene Expression Values (truncated)", 
     xlab="Mean Gene Expression",breaks="scott",xlim=c(2000,600000),ylim = c(0,100))
options(scipen = 999) #turn off scientific notation
```

**Figure 1B. ** Part 1 Question 5: Truncated histogram showing the frequency genes with a mean gene expression over 2000.

***

#### Question 6. Import the “growth_data.csv” file into an R object. What are the column names?

```{r growth_data.csv-codechunk1}
growth.data.URL <- "https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/growth_data.csv" #link to file
growth.data.csv <- read.csv(growth.data.URL) #read in file
colnames(growth.data.csv) #print column names
```

***

#### Question 7. Calculate the mean and standard deviation of tree circumference at the start and end of the study at both sites.


```{r growth_data.csv-codechunk2}
NEsite <- growth.data.csv[grep("northeast", growth.data.csv$Site), ] #grep for text matching, subset for each site
SWsite <- growth.data.csv[grep("southwest", growth.data.csv$Site), ]

bothsites <- list(NEsite$Circumf_2005_cm,NEsite$Circumf_2020_cm,SWsite$Circumf_2005_cm,SWsite$Circumf_2020_cm) #create a list with data from each site
names(bothsites) <- c("NEsite_2005", "NEsite_2020", "SWsite_2005", "SWsite_2020") #2005 = start of study #2020 = end of study

GDmeansd <- function (x) { #function with one input
  GDmean <- signif(mean(x),3)
  GDsd <- signif(sd(x),3)
  paste("Mean = ", GDmean , " SD = ", GDsd)
} #simultaneously calculate mean and SD of x. paste values to user when function is executed

lapply(X=bothsites, FUN = GDmeansd) #execute function on all items in list
```

***

#### Question 8. Make a box plot of tree circumference at the start and end of the study at both sites.


```{r growth_data.csv-codechunk3}
boxplot(NEsite$Circumf_2005_cm,  SWsite$Circumf_2005_cm, 
        NEsite$Circumf_2020_cm, SWsite$Circumf_2020_cm,
        names=c("Northeast 2005", "Southwest 2005", "Northeast 2020", "Southwest 2020"),
        ylab="Tree Circumference (cm)",ylim=c(0,120))
grid() #overlay grid onto figure 
```

**Figure 2. ** Box plot showing tree circumference (cm) at the start (2005) and end (2020) of observations at both the northeast and southwest sites. 

***

#### Question 9. Calculate the mean growth over the last 10 years at each site.


```{r growth_data.csv-codechunk4}
mean10year <- function (x,y) { #function with two inputs, x = 2020, y = 2010
  growth10 <- x - y #10yeargrowth = $Circumf_2020 - $Circumf_2010
  mean10 <- mean(growth10) #mean 10yeargrowth of each
  paste("The mean growth over the past ten years was", mean10 , "cm")
}

mean10year(NEsite$Circumf_2020_cm,NEsite$Circumf_2010_cm) #run function on NEsite and SWsite (subset in Q7)
mean10year(SWsite$Circumf_2020_cm,SWsite$Circumf_2010_cm)
```

***

#### Question 10. Use the t.test and wilcox.test functions to estimate the p-value that the 10 year growth is different at the two sites.

```{r growth_data.csv-codechunk5}
NEgrowth <- NEsite$Circumf_2020_cm - NEsite$Circumf_2010_cm #define 10yeargrowth at each site
SWgrowth <- SWsite$Circumf_2020_cm - SWsite$Circumf_2010_cm
  
ttest10year <- t.test(NEgrowth,SWgrowth) #ttest with two inputs (two sites)

if (ttest10year$p.value < 0.05) { #pull out p value
  message("T-TEST: Mean growth is significantly different (p=", signif(ttest10year$p.value,3), ") between the NE and SW sites over the last 10 years.")
} else {
  message("T-TEST: Mean growth is not significantly different (p=", signif(ttest10year$p.value,3), ") between the NE and SW sites over the last 10 years.")
} #if/else command tells the user if the pvalue is statistically significant (threshold p<0.05)

wilcox10year <- wilcox.test(NEgrowth,SWgrowth) #wilcox test, data must be numeric

if (wilcox10year$p.value < 0.05) {
  message("WILCOX-TEST: Mean growth is significantly different (p=", signif(wilcox10year$p.value,3), ") between the NE and SW sites over the last 10 years.")
} else {
  message("WILCOX-TEST: Mean growth is not significant different (p=", signif(wilcox10year$p.value,3), ") between the NE and SW sites over the last 10 years.")
}
```

***


### Part 2
#### Question 1. Download the whole set of coding DNA sequences for E. coli and your organism of interest. How many coding sequences are present in these organisms? How much coding DNA is there in total for these two organisms? Describe any differences between the two organisms.


```{r part2q1}
ecoliK12URL="http://ftp.ensemblgenomes.org/pub/bacteria/release-53/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655_gca_000005845/cds/Escherichia_coli_str_k_12_substr_mg1655_gca_000005845.ASM584v2.cds.all.fa.gz" #link to URL
download.file(ecoliK12URL,destfile = "ecoli_cds.fa.gz") #download file (compressed)
gunzip("ecoli_cds.fa.gz") #unzip file. returns a file in list format. Each element of list is a unique CDS

hanserisURL="http://ftp.ensemblgenomes.org/pub/bacteria/release-53/fasta/bacteria_51_collection/helicobacter_anseris_gca_003364335/cds/Helicobacter_anseris_gca_003364335.ASM336433v1.cds.all.fa.gz"
download.file(hanserisURL,destfile = "hanseris_cds.fa.gz")
gunzip("hanseris_cds.fa.gz")

ecoli_cds <- seqinr::read.fasta("ecoli_cds.fa") #seqinr package used to perform exploratory analyses on seq data
length(ecoli_cds)  #determine the number of CDS present in each organism 

hanseris_cds <- seqinr::read.fasta("hanseris_cds.fa")
length(hanseris_cds)
```

***

#### Question 2. Calculate the length of all coding sequences in these two organisms. Make a boxplot of coding sequence length in these organisms. What is the mean and median coding sequence length of these two organisms? Describe any differences between the two organisms.


```{r part2q2}
ecoli_cdslength <- sapply(X = ecoli_cds , FUN=length) #calculate the length of each element (CDS) in the list of CDS. #save output as a numerical vector
sum(ecoli_cdslength)

hanseris_cdslength <- sapply(X = hanseris_cds , FUN=length)
sum(hanseris_cdslength) #sum requires vectors, does not work on strings

mean(ecoli_cdslength)
median(ecoli_cdslength)

mean(hanseris_cdslength)
median(hanseris_cdslength)

boxplot(ecoli_cdslength,hanseris_cdslength,
        names=c("E.Coli K-12 (GCA_000005845)", "H.Anseris (GCA_003364335)"),
        ylab="Coding Sequence Length (bp)")
grid() #overlay grid onto figure
```

**Figure 3. ** Box plot depicting the coding sequence length (bp) of E.Coli (gCA_000005845) and H.Anseris (GCA_003364335). Data is shown as median and interquartile range, with whiskers representing the upper and lower quartiles. Outliers are represented as individual points.


***

#### Question 3. Calculate the frequency of DNA bases in the total coding sequences for both organisms (Q3A). Perform the same calculation for the total protein sequence (Q3B). Create bar plots for nucleotide and amino acid frequency. Describe any differences between the two organisms.

**Q3A: **

```{r part2q3a}
ecoli_dna <- unlist(ecoli_cds) #count requires vector, not list. #unlist() to convert list of CDS to character strings
hanseris_dna <- unlist(hanseris_cds)

ecoli_ntfreq <- count(ecoli_dna,1) #count nt (a, g, c, t) frequency #wordsize=single letter = 1
hanseris_ntfreq <- count(hanseris_dna,1)

barplot(as.matrix(rbind(ecoli_ntfreq, hanseris_ntfreq)), beside=TRUE, ylab="frequency", ylim = c(0,1200000), xlab = "nucleotide", main = "E. Coli vs H.Anseris CDS composition",col = c("#009999","#FF9933")) #as.matrix(rbind(x,y)) and beside=TRUE to include both df on same graph
legend("topright", 
       legend = c("E.Coli", "H.Anseris"), 
       fill = c("#009999","#FF9933")) 
grid()

```

**Figure 4A. ** Bar graph depicting nucleotide (A, C, G or T) frequency in the total E.Coli and H.Anseris CDS. 


```{r part2q3a-2}
ecoli_ntprop <- ecoli_ntfreq/sum(ecoli_ntfreq)
hanseris_ntprop <- hanseris_ntfreq/sum(hanseris_ntfreq)

barplot(as.matrix(rbind(ecoli_ntprop, hanseris_ntprop)), beside=TRUE, ylab="nucleotide proportion", ylim = c(0,0.50), 
        xlab = "nucleotide", main = "E. Coli vs H.Anseris CDS composition",col = c("#009999","#FF9933"))
legend("topright", 
       legend = c("E.Coli", "H.Anseri"), 
       fill = c("#009999","#FF9933"))
grid()
```

**Figure 4B. ** Bar graph depicting nucleotide (A, C, G or T) proportion in the total E.Coli and H.Anseris CDS. 


**Q3B: **


```{r part2q3}
ecoli_prot <- lapply(ecoli_cds,translate) #translate CDS into amino acid seq
hanseris_prot <- lapply(hanseris_cds,translate)

aminoacids <- unique(ecoli_prot[[2]]) #define amino acid alphabet using a cDS that contains all 20 amino acids
aminoacids <- aminoacids[aminoacids != "*"] #remove stop codon

ecoli_prots <- unlist(ecoli_prot) #unlist prior to using count function
hanseris_prots <- unlist(hanseris_prot)

ecoli_protscounts <- count(ecoli_prots,wordsize = 1, alphabet = aminoacids) #freq of each amino acid
hanseris_protscounts <- count(hanseris_prots,wordsize = 1, alphabet = aminoacids)

barplot(as.matrix(rbind(ecoli_protscounts, hanseris_protscounts)), beside=TRUE, ylab="frequency", 
        xlab = "amino acid", main = "E. Coli vs H.Anseris protein composition by amino acid",col=c("#009999","#FF9933"))
legend("topright", 
       legend = c("E.Coli", "H.Anseri"), 
       fill = c("#009999","#FF9933"))
grid()
```

**Figure 5A. ** Bar graph depicting amino acid (n=20, stop codon omitted) frequency in the total E.Coli and H.Anseris CDS. 


```{r part2q3-2}
ecoli_protsprop <- ecoli_protscounts/sum(ecoli_protscounts) #normalised to the sum of all amino acount counts 
hanseris_protsprop <- hanseris_protscounts/sum(hanseris_protscounts)

barplot(as.matrix(rbind(ecoli_protsprop, hanseris_protsprop)), beside=TRUE, ylab="proportion", 
        xlab = "amino acid", main = "E. Coli vs H.Anseris protein composition by amino acid",ylim = c(0,0.12),col=c("#009999","#FF9933"))
legend("topright", 
       legend = c("E.Coli", "H.Anseri"), 
       fill = c("#009999","#FF9933"))
grid()
```

**Figure 5B. ** Bar graph depicting amino acid (n=20, stop codon omitted) proportion in the total E.Coli and H.Anseris CDS. 


***

#### Question 4. Create a codon usage table and quantify the codon usage bias among all coding sequences. Describe any differences between the two organisms with respect to their codon usage bias. Provide charts to support your observations.

```{r part2q4}
ecoli_uco <- uco(ecoli_dna,index = "rscu",as.data.frame = TRUE) #uco for codon usage #as.data.frame=TRUE returns indices as a data frame, not vector
ecoli_uco_ordered <- ecoli_uco[order(ecoli_uco$RSCU),]  #order from lowest to highest by RSCU

hanseris_uco <- uco(hanseris_dna,index = "rscu",as.data.frame = TRUE)
hanseris_uco_ordered <- hanseris_uco[order(hanseris_uco$RSCU),]  

ecoli_uco_ordered[c(1:6,59:64),c(1,2,5)] %>% #1:6 lowest 6 by RSCU #59:64 top 6 by RSCU  #pipe to kableExtra
  kbl(caption = "Table 4. Part 2 Question 4: E.Coli. Top six codons used least frequently (teal shading) and top six codons used more frequently (white) than expected (by RSCU).") %>%
  kable_classic(full_width = F, html_font = "Cambria", position = "center", font_size = 14) %>%
  row_spec(1:6, color = "white", background = "#006666") %>% #shading to distinguish between top/bottom six
  row_spec(7:12, color = "#009999", background = "White" )

hanseris_uco_ordered[c(1:6,59:64),c(1,2,5)] %>% 
  kbl(caption = "Table 5. Part 2 Question 4: H.Anseris. Top six codons used least frequently (orange shading) and top six codons used more frequently (white) than expected (by RSCU).") %>%
  kable_classic(full_width = F, html_font = "Cambria", position = "center", font_size = 14) %>%
  row_spec(1:6, color = "white", background = "#FF9933") %>% 
  row_spec(7:12, color = "#FF9933", background = "White" )
```

***

#### Question 5. In the organism of interest, identify 10 protein sequence k-mers of length 3-5 which are the most over- and under-represented k-mers in your organism of interest. Are these k-mers also over- and under-represented in E. coli to a similar extent? Provide plots to support your observations. Why do you think these sequences are present at different levels in the genomes of these organisms?

```{r part2q5}
ecoli_3merfreq <- count(ecoli_prots,wordsize=3,alphabet=aminoacids,freq=TRUE) #3-mer amino acid sequence by frequency
ecoli_3merfreq_ordered <- ecoli_3merfreq[order(ecoli_3merfreq)] #order, lowest to highest as default

ecoli_4merfreq <- count(ecoli_prots,wordsize=4,alphabet=aminoacids,freq=TRUE) #4-mer amino acid sequence by frequency
ecoli_4merfreq_ordered <- ecoli_4merfreq[order(ecoli_4merfreq)] #order, lowest to highest as default

ecoli_5merfreq <- count(ecoli_prots,wordsize=5,alphabet=aminoacids,freq=TRUE) #4-mer amino acid sequence by frequency
ecoli_5merfreq_ordered <- ecoli_5merfreq[order(ecoli_5merfreq)] #order, lowest to highest as default

#pipe 1:10 (10 underrep) and last 10 (use nrow to determine max row) to kableExtra to present
ecoli_3merfreq_ordered[c(1:10,7991:8000)] %>% 
  kbl(caption = "Table 6A. Part 2 Question 5: E.Coli. 10 under-represented (teal shading) over-represented (white) amino acid 3-mers") %>%
  kable_classic(full_width = F, html_font = "Cambria", position = "center", font_size = 14) %>%
  row_spec(1:10, color = "white", background = "#006666") %>% #shading to distinguish between 10 over and under- rep
  row_spec(11:20, color = "#009999", background = "White" )

ecoli_4merfreq_ordered[c(1:10,159991:160000)] %>% #1:6 lowest 6 by RSCU #59:64 top 6 by RSCU  #pipe to kableExtra
  kbl(caption = "Table 6B. Part 2 Question 5: E.Coli. 10 under-represented (teal shading) over-represented (white) amino acid 4-mers") %>%
  kable_classic(full_width = F, html_font = "Cambria", position = "center", font_size = 14) %>%
  row_spec(1:10, color = "white", background = "#006666") %>% 
  row_spec(11:20, color = "#009999", background = "White" )

ecoli_5merfreq_ordered[c(1:10,3199991:3200000)] %>% 
  kbl(caption = "Table 6C. Part 2 Question 5: E.Coli. 10 under-represented (teal shading) over-represented (white) amino acid 5-mers") %>%
  kable_classic(full_width = F, html_font = "Cambria", position = "center", font_size = 14) %>%
  row_spec(1:10, color = "white", background = "#006666") %>% 
  row_spec(11:20, color = "#009999", background = "White") 


#now repeated for 3-, 4- and 5-mers using the hanseris_prots sequence

hanseris_3merfreq <- count(hanseris_prots,wordsize=3,alphabet=aminoacids,freq=TRUE)
hanseris_3merfreq_ordered <- hanseris_3merfreq[order(hanseris_3merfreq)] 

hanseris_4merfreq <- count(hanseris_prots,wordsize=4,alphabet=aminoacids,freq=TRUE)
hanseris_4merfreq_ordered <- hanseris_4merfreq[order(hanseris_4merfreq)] 

hanseris_5merfreq <- count(hanseris_prots,wordsize=5,alphabet=aminoacids,freq=TRUE) 
hanseris_5merfreq_ordered <- hanseris_5merfreq[order(hanseris_5merfreq)] 

#pipe 1:10 (10 underrep) and last 10 (use nrow to determine max row) to kableExtra to present
hanseris_3merfreq_ordered[c(1:10,7991:8000)] %>% 
  kbl(caption = "Table 7A. Part 2 Question 5: H.Anseris. 10 under-represented (teal shading) over-represented (white) amino acid 3-mers") %>%
  kable_classic(full_width = F, html_font = "Cambria", position = "center", font_size = 14) %>%
  row_spec(1:10, color = "white", background = "#FF9933") %>% #shading to distinguish between 10 over and under- rep
  row_spec(11:20, color = "#FF9933", background = "White" )

hanseris_4merfreq_ordered[c(1:10,159991:160000)] %>% #1:6 lowest 6 by RSCU #59:64 top 6 by RSCU  #pipe to kableExtra
  kbl(caption = "Table 7B. Part 2 Question 5: H.Anseris. 10 under-represented (teal shading) over-represented (white) amino acid 4-mers") %>%
  kable_classic(full_width = F, html_font = "Cambria", position = "center", font_size = 14) %>%
  row_spec(1:10, color = "white", background = "#FF99336") %>% 
  row_spec(11:20, color = "#FF9933", background = "White" )

hanseris_4merfreq_ordered[c(1:10,3199991:3200000)] %>% 
  kbl(caption = "Table 7C. Part 2 Question 5: H.Anseris. 10 under-represented (teal shading) over-represented (white) amino acid 5-mers") %>%
  kable_classic(full_width = F, html_font = "Cambria", position = "center", font_size = 14) %>%
  row_spec(1:10, color = "white", background = "#FF9933") %>% 
  row_spec(11:20, color = "#FF9933", background = "White" )
```

***

```{r sessionInfo}
sessionInfo()
```
