---
title: 'SLE712 AT3: BIoinformatics Assignment'
author: "Jessica Silver"
date: '2022-05-07'
output: html_document
---

source: LINK TO GITHUB REPO

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

#### Set Up Instructions 

Include some details here about packages that must be installed prior to running these codes. This information may also be included in the readme 



```{r setupinstructions}
suppressPackageStartupMessages({
  library("seqinr")
library("kableExtra")
library("R.utils")
})

```

### Part 1
#### Question 1. Read in the file, making the gene identifiers the row names. Show a table of values for the first six genes. 

The link to the file (URL) was first defined as "gene.expression.URL". The table was then read in to Rstudio using the `read.table` command. 

The structure of the file was first viewed using `str(df)`. 

As a default, integers were given as row and column names. 

`header=TRUE` and `row.names = 1` were included in the `read.table(df)` command to make the gene identifiers the row names. The object was named "gene.expression.tsv".

A table for values for the first six genes was generated using the `df[row,columns]` command structure. As all columns for the first 6 genes (rows) were requested, the command reads `gene.expression.tsv[1:6,]` **CHANGE THIS: USING kable AND APPROPRIATE COMMANDS. 


``` {r gene.expression.tsv-codechunk1}
gene.expression.URL <- "https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/gene_expression.tsv"
gene.expression.tsv <- read.table(gene.expression.URL,stringsAsFactors = FALSE, header=TRUE, row.names = 1)

gene.expression.tsv[1:6,]
```
**Table 1.** Part 1 Question 1. Table of values for the first six genes.



#### Question 2. Make a new column which is the mean of the other columns. Show a table of values for the first six genes.
`rowmeans(df)` was used to calculate average gene expression between the three samples of interest for all genes. 

A new column containing the average gene expression was added to the "gene.expression.tsv" dataframe using `df$newcolumnname` notation. 

A table for values for the first six genes was generated using the `df[row,columns]` command structure as described above. 

``` {r gene.expression.tsv-codechunk2}
gene.expression.tsv$GE.Means <- rowMeans(gene.expression.tsv)
gene.expression.tsv[1:6,]
```
**Table 2.** Part 1 Question 2. Table of values for the first six genes, including new "Mean Gene Expression" Column. 



#### Question 3. List the 10 genes with the highest mean expression
`order()` cmd sorts from lowest to highest numerical value as a default. Therefore, the lowest gene expression values will be presented first. 
As we want to report the 10 genes with the highest mean expression, the command was modified to `order(-)` to reverse sort the data from highest to lowest. 
`df[row,columns]` format used. Question asks to sort the genes, which are the row names. Therefore `order(-)` is placed in the first part of `[]` and columns section is left blank. 

The sorted dataset was named "reverse_sortedGEmeans"

`df[row,column]` was executed to show a specific part of the dataframe. 
`[1:10,n]` shows the first 10 rows corresponding to 10 highest genes
`[n,4]` 4 corresponds to the fourth column in the df = Mean Gene Expression
`[n,n,drop=FALSE]` to retain the dimesions of the object.
 

``` {r gene.expression.tsv-codechunk3}
reverse_sortedGEMeans <- gene.expression.tsv[order(-gene.expression.tsv$GE.Means),]
reverse_sortedGEMeans[1:10,4,drop=FALSE] #drop=FALSE retains the dimension of the object
```
**Table 3.** Part 1 Question 3. Table showing the top 10 genes with the highest mean expression across the three samples. 

#### Question 4. Determine the number of genes with a mean < 10
`nrow()` is the base command used to return the number of genes (i.e rows) in the dataframe.

The number of genes with a mean < 10 was determined using the `df[row,column]` command and embedded into `nrow`.

`print` may be used to return the number of genes with a mean less than 10. 


```{r gene.expression.tsv-codechunk4}
nrow(gene.expression.tsv[gene.expression.tsv$GE.Means <10,])
print(nrow(gene.expression.tsv[gene.expression.tsv$GE.Means <10,]))
```

#### Question 5. Make a histogram plot of the mean values and include it into your report.
`hist()` command used to plot a histogram of the mean values. 
The `df$columnname` was specified, header included using `main=""` and x axis title included using `xlab=""`. 

x and y axis limits were specified using `xlim=()` and `ylim=()`, respectively.

```{r gene.expression.tsv-codechunk5}
hist(gene.expression.tsv$GE.Means,main="Part-1 Q-5: Mean Gene Expression Values", xlab="Mean Gene Expression",breaks="scott",xlim=c(0,10000),ylim = c(0,50000))
```

**Table 4.** Part 1 Question 5. 


#### Question 6. Import the “growth_data.csv” file into an R object. What are the column names?
The link to the file (URL) was first defined as "growth.data.URL". The csv file was then read in to Rstudio using the `read.csv` command and named "growth.data.csv"

`colnames(df)` was used to extract the column names, and prints "Site", "Tree ID", "Circumf_2005_cm", ", Circumf_2010_cm", "Circumf_2015_cm" and "Circumf_2020_cm"

```{r growth_data.csv-codechunk1}
growth.data.URL <- "https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/growth_data.csv"
growth.data.csv <- read.csv(growth.data.URL)
colnames(growth.data.csv)
```

#### Question 7. Calculate the mean and standard deviation of tree circumference at the start and end of the study at both sites.

From the column names, the start of the study is "Circumf_2005_cm" (2005) and the end of the stuydy is "Circumf_2020_cm" (2020). There are two sites named "northeast" and "southwest". 

`grep` was used to create new separate objects for each site ("NEsite" and "SWsite").
A `list()` was created containing ... 

A new function was defined ("GDmeansd") that requires one input and will print the mean and SD of the input. 

`lapply` was used to calculate the mean and SD growth of trees at each site at the start and end of each in one command. 

```{r growth_data.csv-codechunk2}
NEsite <- growth.data.csv[grep("northeast", growth.data.csv$Site), ]
SWsite <- growth.data.csv[grep("southwest", growth.data.csv$Site), ]

bothsites <- list(NEsite$Circumf_2005_cm,NEsite$Circumf_2020_cm,SWsite$Circumf_2005_cm,SWsite$Circumf_2020_cm)
names(bothsites) <- c("NEsite_2005", "NEsite_2020", "SWsite_2005", "SWsite_2020")

GDmeansd <- function (x) {
  GDmean <- signif(mean(x),3)
  GDsd <- signif(sd(x),3)
  paste("Mean = ", GDmean , " SD = ", GDsd)
}

lapply(X=bothsites, FUN = GDmeansd)
```


#### Question 8. Make a box plot of tree circumference at the start and end of the study at both sites.

`boxplot()` used to generate a box plot. 

```{r growth_data.csv-codechunk3}
boxplot(NEsite$Circumf_2005_cm,  SWsite$Circumf_2005_cm, 
        NEsite$Circumf_2020_cm, SWsite$Circumf_2020_cm,
        names=c("Northeast 2005", "Southwest 2005", "Northeast 2020", "Southwest 2020"),
        ylab="Tree Circumference (cm)",ylim=c(0,120))
grid()
```

**Figure 1** Box plot showing tree circumference (cm) at the start (2005) and end (2020) of observations at both the northeast and southwest sites. 


#### Question 9. Calculate the mean growth over the last 10 years at each site.

A `function` with two inputs `(x,y)` was used to first calculate the growth over the last 10 years at each site. 

10 years = `$Circumf_2020 - $Circumf_2010` 

Next, `mean()` command was used to calculate the average 10 year growth each site. 

The function "mean10year" was then run on the previously defined "NEsite" and "SWsite" subsets. 

The output of the function prints a message including the mean growth of all trees at each site over the past 10 years. 

**ANSWER** 

NE site = 42.94cm

SW site = 35.49cm

```{r growth_data.csv-codechunk4}
mean10year <- function (x,y) { 
  growth10 <- x - y 
  mean10 <- mean(growth10)
  paste("The mean growth over the past ten years was", mean10 , "cm")
}

mean10year(NEsite$Circumf_2020_cm,NEsite$Circumf_2010_cm)
mean10year(SWsite$Circumf_2020_cm,SWsite$Circumf_2010_cm)
```

#### Question 10. Use the t.test and wilcox.test functions to estimate the p-value that the 10 year growth is different at the two sites.

10 year growth from the NE and SW sites (`$Circumf_2020 - $Circumf_2010`) was carried forward from question 9. 

`t.test()` command with two inputs was performed to estimate the p value of the 10 year growth between each site. The output was saved as a new object ("ttest10year"). The p value from this ttest was printed to the user using the `print` function.

An `if/else` command was incorporated to tell the user if the p value is statistically significant (p<0.05). 

```{r growth_data.csv-codechunk5}
NEgrowth <- NEsite$Circumf_2020_cm - NEsite$Circumf_2010_cm
SWgrowth <- SWsite$Circumf_2020_cm - SWsite$Circumf_2010_cm
  
ttest10year <- t.test(NEgrowth,SWgrowth)
print(signif(ttest10year$p.value,3))

if (ttest10year$p.value < 0.05) {
  message("Mean growth is significantly different (p=", signif(ttest10year$p.value,3), ") between the NE and SW sites over the last 10 years.")
} else {
  message("Mean growth is not significantly different (p>0.05) between the NE and SW sites over the last 10 years.")
}
```


**make sure to include session info before submitting. this must include kable, and other packages used in week10**